### ISSUE TITLE
uftrace info enhancement for better task display

### ISSUE THREAD

**Issue by honggyukim** [namhyung/uftrace#645]
Some target programs have complicated process and thread creation model.

In this case, it'd be good to have a way to see their relations in a better way with following features:
- Display task info in tree format just like `uftrace graph` output.
  - The usage may be `uftrace info --task-tree`
- Distinguish process and thread in `task list` output.
  - We may put `*` to process or thread or any other way.

**Comment by honggyukim** [comment]
It'd be really useful when analyzing big programs such as `chrome`, which may have more than 40 tasks in a single record.

**Comment by honggyukim** [comment]
We can also add the lifetime of each task because many tasks are very close to have empty trace records.

**Comment by namhyung** [comment]
We can reuse `--threads` option IMO.  Also perf event has task fork/exit info so it should be enough to build task tree with lifetime.

**Comment by namhyung** [comment]
Please test [check/info-update](https://github.com/namhyung/uftrace/compare/check/info-update).  I just reused `--threads` option, maybe I think it's better to rename it as `--task`.  Note that threads will be indented with 3 spaces and child will be with 6.

```
$ uftrace info --threads
 11.307 ms [255272] parent
  4.039 ms       [255274] child
```

**Comment by honggyukim** [comment]
Thanks for the work! It looks good but it doesn't show updated task name for chrome.
```
  9.002  m [168086] chrome
              [168130] chrome
  7.018  m       [168131] chrome
  2.039  m             [168200] chrome
                          [168201] chrome
  2.039  m                [168202] chrome
  2.039  m                [168204] chrome
  2.039  m                [168203] chrome
  2.039  m                [168205] chrome
  2.039  m                [168206] chrome
  2.039  m                [168207] chrome
  2.039  m                [168208] chrome
  2.039  m                [168209] chrome
  2.039  m                [168210] chrome
  2.039  m                [168211] chrome
                          [168212] chrome
                          [168331] chrome
  2.037  m             [168255] chrome
                          [168257] chrome
    ...
  3.038  m    [168333] chrome
  3.036  m    [168334] chrome
  3.031  m       [168344] chrome
  2.013  m          [168441] chrome
  2.013  m          [168442] chrome
  2.013  m          [168443] chrome
  2.013  m          [168444] chrome
 33.000  s          [168445] chrome
 33.117  s          [168446] chrome
  2.010  m          [168455] chrome
 33.000  s          [168637] chrome
                    [168674] chrome
  3.002  m    [168399] chrome
  1.006  m    [168398] chrome
```
In addition, I think it'd be really better to have the following features:
- display it same as `graph` command.
- add a new field "task tree" in `tui` so that we can fold and unfold just like call graph.
- apply time filter based on it's life time.

The above requests are optional so no need to do them all. Thanks a lot for your effort.

**Comment by honggyukim** [comment]
For your reference, I will attach the data recorded for chrome.
[uftrace.data.youtube.google.naver.debug.tar.gz](https://github.com/namhyung/uftrace/files/3051045/uftrace.data.youtube.google.naver.debug.tar.gz)

**Comment by honggyukim** [comment]
It would be also needed to filter out some small tasks based on time-filter even in `uftrace info` output.
```
$ uftrace info --no-pager
    ...
# process information
# ===================
# number of tasks     : 514
# task list           : 29022(chrome), 29037(sandbox_ipc_thr), 29038(chrome), 29042(Chrome_IOThread), 
29053(NetworkChangeNo), 29054(D-Bus thread), 29055(CrShutdownDetec), 29056(inotify_reader), 
29057(TaskSchedulerSe), 29058(TaskSchedulerFo), 29059(TaskSchedulerFo), 
29060(TaskSchedulerFo), 29061(TaskSchedulerFo), 29062(TaskSchedulerFo),
29063(TaskSchedulerFo), 29064(TaskSchedulerFo), 29065(TaskSchedulerFo),
29066(TaskSchedulerFo), 29067(TaskSchedulerFo), 29068(TaskSchedulerSi), 29069(CompositorTileW),
29070(AudioThread), 29071(TaskSchedulerSi), 29072(gpu-process_cra), 29073(chrome),
29074(CacheThread_Blo), 29075(BrowserWatchdog), 29076(xdg-settings), 29077(bash), 29078(dbus-send),
29079(xprop), 29080(grep), 29081(xprop), 29082(grep), 29083(bash), 29084(grep), 29085(bash), 
29086(uname), 29087(bash), 29088(xdg-mime), 29089(bash), 29090(dbus-send), 29091(xprop), 
29092(grep), 29093(xprop), 29094(grep), 29098(bash), 29099(grep), 29100(bash), 29101(uname), 
29102(bash), 29103(bash), 29104(sed), 29105(bash), 29106(grep), 29108(head), 29109(cut), 
29110(cut), 29111(bash), 29112(grep), 29113(head), 29114(cut), 29115(cut), 29116(bash), 29117(grep), 
29118(head), 29119(cut), 29120(cut), 29121(bash), 29122(grep), 29123(head), 29124(cut), 29125(cut), 
29126(bash), 29127(grep), 29128(head), 29129(cut), 29130(cut), 29133(bash), 29134(bash), 
29135(which), 29136(bash), 29137(readlink), 29138(bash), 29139(basename), 29140(grep), 
    ...
29624(TaskSchedulerFo), 29625(TaskSchedulerFo), 29626(TaskSchedulerFo), 
29627(TaskSchedulerFo), 29628(BatteryStatusNo), 29629(TaskSchedulerSi), 29630(Media), 
29631(chrome), 29632(AudioOutputDevi), 29633(ServiceWorker t), 29634(Shutdown watchd), 
29663(chrome)
# exe image           : /mnt/data1/home3/honggyu/work/chromium/neva/chromium/src/out/Release.pg/chrome
# auto-args           : true
# pattern             : regex
# exit status         : exited with code: 0
# elapsed time        : 109.437285730 sec
# cpu time            : 2.816 / 61.995 sec (sys / user)
# context switch      : 23722 / 394 (voluntary / involuntary)
# max rss             : 242044 KB
# page fault          : 1 / 383071 (major / minor)
# disk iops           : 32 / 9432 (read / write)
```
It has too much small tasks with just a simple task such as `grep`, `cut`, `head`, etc. I'm not interested those tasks at all so it'd be great if we can filter-out with time-filter and only shows the reduced number of tasks.

**Comment by honggyukim** [comment]
I think it'd be really good to show the idle time (or running time) of each task as well using schedule event. If it doesn't have schedule event, then it only shows `TASK TIME` for lifetime of tasks.
```
#  TASK TIME   IDLE TIME   TASK
    2.755 ms             : (8385) task1
    1.897 us    1.897 us :  +-(8396) task2
                         :  |
    1.356 us    1.356 us :  +-(8412) task3
                         :  |
    2.752 ms   34.621 us :  +-(8413) task4
  446.888 us  446.888 us :     +-(8421) task5
                         :     |
    2.257 ms   26.590 us :     +-(8425) task6
    2.230 ms    2.230 ms :     | (8426) task7
                         :     |
   22.334 us    2.084 us :     +-(8474) task8
```

**Comment by honggyukim** [comment]
It'd be also beneficial if we can put `*` for processes to distinguish them from threads.

**Comment by honggyukim** [comment]
Thinking it again. We don't have to use this feature to `info` command. The above output can be displayed with `uftrace graph --threads`.

In addition to this, we can also add `uftrace replay --threads` as well to show their lifetime in timeline based sequence.

I also think that we better rename it from `--threads` to `--task`

**Comment by namhyung** [comment]
I also think that `graph` might be a better choice but either is fine to me.  Adding idle time looks like a good idea.

**Comment by honggyukim** [comment]
Enhancing this task graph info can be very useful to show the big picture to show what's going on inside the program.

Sorry for adding up too many things here, but I always curious "WHERE" and "HOW" the each task is created. This graph shows "WHO" created the task, but if we have a backtrace info when the new task is created, that would provide very useful info.

**Comment by namhyung** [comment]
Your suggestion is always good and very helpful. :)

**Comment by honggyukim** [comment]
I appreciate all your support as well. Thanks! :)

**Comment by namhyung** [comment]
We may enhance `uftrace report --threads` as well.

**Comment by honggyukim** [comment]
Sure. I also think so and would like to expand it to maybe `replay` command.

It can show what tasks are currently executed and how the program runs in parallel. It could also show thread level contention without detailed function level info.

**Comment by honggyukim** [comment]
As mentioned in #646, I think it'd be better to add another option called `--taskinfo` to show task-level info rather than rename it to `--task`.

We can use `--threads` to work for each tids when it's used with analysis commands - `replay`, `report`, and `graph`. Maybe we better change `report --threads` output here and replace its usage with `--taskinfo`.

**Comment by namhyung** [comment]
Why not just using a single option (`--task`) like below?

* `replay --task` shows per-task output sequentially
* `report --task` shows task level execution summary (with time filter)
* `graph --task` shows task relationship

**Comment by honggyukim** [comment]
We can use `--task` for task-level analysis for each command. The existing `--threads` can be used still function-level analysis but it can show the output for each task separately just like we show graph output for each session one by one.

The current replay output for `t-fork` is as follows:
```
$ uftrace replay
# DURATION     TID     FUNCTION
   1.643 us [ 73051] | __monstartup();
   1.277 us [ 73051] | __cxa_atexit();
            [ 73051] | main() {
 197.341 us [ 73051] |   fork();
            [ 73051] |   wait() {
            [ 73064] |   } /* fork */
            [ 73064] |   a() {
            [ 73064] |     b() {
            [ 73064] |       c() {
   1.554 us [ 73064] |         getpid();
   2.827 us [ 73064] |       } /* c */
   3.310 us [ 73064] |     } /* b */
   8.267 us [ 73064] |   } /* a */
  18.150 us [ 73064] | } /* main */
 775.974 us [ 73051] |   } /* wait */
            [ 73051] |   a() {
            [ 73051] |     b() {
            [ 73051] |       c() {
   2.404 us [ 73051] |         getpid();
   3.440 us [ 73051] |       } /* c */
   3.890 us [ 73051] |     } /* b */
   8.541 us [ 73051] |   } /* a */
   1.003 ms [ 73051] | } /* main */
```
I mean to use `--threads` to show replay output for each task one by one.
```
$ uftrace replay --threads
# Function Replay for tid 73051
# DURATION     TID     FUNCTION
   1.643 us [ 73051] | __monstartup();
   1.277 us [ 73051] | __cxa_atexit();
            [ 73051] | main() {
 197.341 us [ 73051] |   fork();
 775.974 us [ 73051] |   wait();
            [ 73051] |   a() {
            [ 73051] |     b() {
            [ 73051] |       c() {
   2.404 us [ 73051] |         getpid();
   3.440 us [ 73051] |       } /* c */
   3.890 us [ 73051] |     } /* b */
   8.541 us [ 73051] |   } /* a */
   1.003 ms [ 73051] | } /* main */

# Function Replay for tid 73064
# DURATION     TID     FUNCTION
            [ 73064] | } /* fork */
            [ 73064] | a() {
            [ 73064] |   b() {
            [ 73064] |     c() {
   1.554 us [ 73064] |       getpid();
   2.827 us [ 73064] |     } /* c */
   3.310 us [ 73064] |   } /* b */
   8.267 us [ 73064] | } /* a */
  18.150 us [ 73064] | } /* main */
```
The same way can be applied to report and graph commands.

**Comment by honggyukim** [comment]
So we now have two different view of recorded data with `--task` option.
1. function-level output
2. task-level output

In addition to `--task` option, I would like to provide a way that doesn't mix the output with multiple tids with `--threads` option. I feel it's very confuse to see (especially) replay output with different depth.

I would like to specially hide worker threads in replay output because they cannot be hidden by time filter because they have a long duration although they are mostly in idle status. If we have a way to separately see each task one by one, it's very easy to see them out of scope from the main thread.

**Comment by namhyung** [comment]
I don't get what you want clearly.  Could you please show me an example output of the two?

**Comment by honggyukim** [comment]
As I wrote above, the replay output with `--threads` is still function-level view but shows each task one by one. It looks as follows:
```
$ uftrace replay --threads
# Function Replay for tid 73051
# DURATION     TID     FUNCTION
   1.643 us [ 73051] | __monstartup();
   1.277 us [ 73051] | __cxa_atexit();
            [ 73051] | main() {
 197.341 us [ 73051] |   fork();
 775.974 us [ 73051] |   wait();
            [ 73051] |   a() {
            [ 73051] |     b() {
            [ 73051] |       c() {
   2.404 us [ 73051] |         getpid();
   3.440 us [ 73051] |       } /* c */
   3.890 us [ 73051] |     } /* b */
   8.541 us [ 73051] |   } /* a */
   1.003 ms [ 73051] | } /* main */

# Function Replay for tid 73064
# DURATION     TID     FUNCTION
            [ 73064] | } /* fork */
            [ 73064] | a() {
            [ 73064] |   b() {
            [ 73064] |     c() {
   1.554 us [ 73064] |       getpid();
   2.827 us [ 73064] |     } /* c */
   3.310 us [ 73064] |   } /* b */
   8.267 us [ 73064] | } /* a */
  18.150 us [ 73064] | } /* main */
```
However, if we use `--task` or `--taskinfo`, then its output mode is task-level view and shows the timeline based replay output but each element is a task begin or end, not a function entry or exit.

It can be displayed similar to [Gantt chart](https://en.wikipedia.org/wiki/Gantt_chart).

The following output looks very sloppy, but may show the idea. I would like to show the lifetime  of each task.
```
#                TIDS
# DURATION      100860      100871
#               t-fork      t-fork
 339.125 us |     ##
  11.427 us |                 ##
   1.288 ms |     ##
```
In `graph` command with `--task`, then the output looks as follows:
```
#  TASK TIME   IDLE TIME   TASK
    2.755 ms             : (8385) task1
    1.897 us    1.897 us :  +-(8396) task2
                         :  |
    1.356 us    1.356 us :  +-(8412) task3
                         :  |
    2.752 ms   34.621 us :  +-(8413) task4
  446.888 us  446.888 us :     +-(8421) task5
                         :     |
    2.257 ms   26.590 us :     +-(8425) task6
    2.230 ms    2.230 ms :     | (8426) task7
                         :     |
   22.334 us    2.084 us :     +-(8474) task8
```
But it can also show function-level view, but each task one by one.

**Comment by namhyung** [comment]
The `replay --task` above looks nice but I'm not sure it's ok for the case where many tasks are involved since it'd run out of the columns.

**Comment by honggyukim** [comment]
I agree. It's just a conceptual output but I will think about the better way. Thanks.

**Comment by namhyung** [comment]
Pushed [review/task-info-v1](https://github.com/namhyung/uftrace/compare/review/task-info-v1)

It has some incorrect time, will fix later..

```
$ uftrace graph --task -d uftrace.data.chrome
========== TASK GRAPH ==========
#   RUN-TIME    CPU-TIME : TASK
    9.002  m    1.041  m : [168086] chrome
                         :  +-[168130] chrome
                         :  |    
    7.018  m    1.020  m :  +----[168131] chrome
                         :  |     |    
    2.039  m    2.802  s :  |     +----[168200] chrome
                         :  |     |     +-[168201] chrome
    2.039  m   16.797 ms :  |     |     +-[168202] TaskSchedulerFo
    2.039  m  432.881 ms :  |     |     +-[168204] Chrome_ChildIOT
    2.039  m    4.465 ms :  |     |     +-[168203] TaskSchedulerFo
    2.039  m    1.002 ms :  |     |     +-[168205] GpuMemoryThread
    2.039  m   14.379 ms :  |     |     +-[168206] Font_Proxy_Thre
    2.039  m  384.770 ms :  |     |     +-[168207] Compositor
    2.039  m   60.963 ms :  |     |     +-[168208] CompositorTileW
    2.039  m   45.158 ms :  |     |     +-[168209] CompositorTileW
    2.039  m   44.586 ms :  |     |     +-[168210] CompositorTileW
    2.039  m   49.438 ms :  |     |     +-[168211] CompositorTileW
                         :  |     |     +-[168212] chrome
                         :  |     |     +-[168331] chrome
                         :  |     |    
    2.037  m  242.418 ms :  |     +----[168255] chrome
                         :  |     |     +-[168257] chrome
    2.037  m  742.956 us :  |     |     +-[168258] TaskSchedulerFo
    2.037  m    1.977 ms :  |     |     +-[168259] TaskSchedulerFo
    2.037  m   83.156 ms :  |     |     +-[168260] Chrome_ChildIOT
    2.037  m    1.387 ms :  |     |     +-[168263] GpuMemoryThread
    2.037  m  885.150 us :  |     |     +-[168264] Font_Proxy_Thre
    2.037  m    8.357 ms :  |     |     +-[168268] Compositor
  ...
```

**Comment by honggyukim** [comment]
Looks very good! I appreciate your effort on this.

I have another idea for task view. It looks good to see who created which task.

I have another idea but not sure if it's possible. I will just leave the idea here. I also think that it can also be useful if we can provide task waken-up tree, which shows the opposite view of task creation tree.

**Comment by namhyung** [comment]
Pushed [review/task-info-v2](https://github.com/namhyung/uftrace/compare/review/task-info-v2)

Changelog:
 * add `--task` option to info command
 * update `--task` option behavior of report command
 * support some sort keys for task report

For example:

```
$ uftrace report -task -s runtime
     TID    Run time    CPU time   Num funcs  Task name       
  ======  ==========  ==========  ==========  ================
  168086    7.018  m   17.692  s       30501  chrome
  168131    5.059  m  653.362 ms         194  chrome
  168148    5.058  m    7.183 ms         123  TaskSchedulerSe
  168149    5.058  m   22.952 ms         367  TaskSchedulerFo
  168151    5.058  m    4.441  s        7036  TaskSchedulerFo
  168152    5.058  m   11.533  s       33228  Chrome_IOThread
  168154    5.055  m    5.168 ms          12  NetworkChangeNo
  ...
```

The report command calculate runtime and cputime based on duration of each function.  In some case, it has a bit different result that graph command which calculate it simply based on the first and last timestamp.

```
$ uftrace info --task
#         TIMESTAMP       FLAGS     PID    TASK              DATA SIZE
       1133348.699393850  FS     [168086]  chrome                0.374 MB
       1133452.507146691         [168130]  chrome           
       1133452.529798102  FS     [168131]  chrome                0.003 MB
       1133532.599866572         [168148]  TaskSchedulerSe       0.002 MB
       1133532.600091739         [168149]  TaskSchedulerFo       0.006 MB
       1133532.600451420         [168150]  TaskSchedulerFo       0.002 MB
       1133532.600740387         [168151]  TaskSchedulerFo       0.175 MB
       1133532.601072727         [168152]  Chrome_IOThread       0.553 MB
       1133532.640979853         [168153]  chrome           
       1133534.957256268         [168154]  NetworkChangeNo       0.000 MB
  ...
```

The info command shows first timestamp and data size.  The flags say that 'F' for forked process and 'S' for new session.

**Comment by honggyukim** [comment]
Looks very good!! Thanks for the update again. I just have a few comments on this from the following output.
- It'd be better to switch the columns of `FLAGS` and `PID`
- I also think that `PID` is better to be `TID`.
- Lastly, I'm not sure what kind of information `TIMESTAMP` can deliver to users. What about just print the session id instead so that users can open the session file easily? The session file can also be opened in TUI mode.
```
$ uftrace info --task
#         TIMESTAMP       FLAGS     PID    TASK              DATA SIZE
       1133348.699393850  FS     [168086]  chrome                0.374 MB
       1133452.507146691         [168130]  chrome           
       1133452.529798102  FS     [168131]  chrome                0.003 MB
       1133532.599866572         [168148]  TaskSchedulerSe       0.002 MB
       1133532.600091739         [168149]  TaskSchedulerFo       0.006 MB
       1133532.600451420         [168150]  TaskSchedulerFo       0.002 MB
       1133532.600740387         [168151]  TaskSchedulerFo       0.175 MB
       1133532.601072727         [168152]  Chrome_IOThread       0.553 MB
       1133532.640979853         [168153]  chrome           
       1133534.957256268         [168154]  NetworkChangeNo       0.000 MB
  ...
```
In addition, it seems that the name `Run time` looks a bit vague because it didn't actually run during idle time. IMHO, `Total time` looks better and more consistent to the original report output.

**Comment by namhyung** [comment]
I thought timestamp can be a hint for `--time-range` option.  Also pid and task name tend to be together, will rename to tid anyway.

**Comment by namhyung** [comment]
Pushed [review/task-info-v3](https://github.com/namhyung/uftrace/compare/review/task-info-v3)

Changelog:
 * rename to 'Total/self time' rather than 'run/cpu time'
 * rename 'pid' to 'tid'
 * add a test case

**Comment by honggyukim** [comment]
I just saw this feature is merged into master. Thanks for your hard work!

But I have a few requests to enhance this.
- I think the output can be shown exactly same as normal `graph` output.
- It doesn't work well with time-filter. I expect the time-filter work to hide based on the task time. It could be `total` or `self` time

**Comment by namhyung** [comment]
Please file a new issue for it.
